### PR TITLE
gh-104469: Disallow using Py_LIMITED_API with Py_BUILD_CORE

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-104469

### CODE REVIEW

**Comment by vstinner** [comment]
When I reviewed https://github.com/python/cpython/pull/107857 I noticed that Py_LIMITED_API was used with Py_BUILD_CORE which makes no sense to me. Make a choice: either the limited C API, or the internal C API :-)

**Comment by vstinner** [comment]
@corona10 [wrote](https://github.com/python/cpython/pull/107857#issuecomment-1678182217):
> Okay, this is an issue between AC and limited C APIs. We should revert it first.
> cc @erlend-aasland @vstinner @nahyeon-an

In the meanwhile, AC got support for the limited C API ;-)

File: Tools/c-analyzer/c_parser/preprocessor/gcc.py [hunk 1]
```diff
@@ -3,13 +3,18 @@
 
 from . import common as _common
 
-# The following C files must not be built with Py_BUILD_CORE,
-# because they use the limited C API.
+# The following C files defines the Py_LIMITED_API macro, and so must not be
```
Comment (colesbury) [inline Tools/c-analyzer/c_parser/preprocessor/gcc.py]: "define" (instead of "defines")

Comment (vstinner) [reply Tools/c-analyzer/c_parser/preprocessor/gcc.py]: oops, fixed!

**Comment by vstinner** [comment]
Merged. Thanks for the review @colesbury.

**Comment by erlend-aasland** [comment]
Thanks! LGTM

**Comment by vstinner** [comment]
Oh. This change broke WASM buildbots.

Example with wasm32-emscripten node (dynamic linking) 3.x: https://buildbot.python.org/all/#/builders/1056/builds/3142

```
/opt/emsdk/upstream/emscripten/emcc  -DNDEBUG -g -O3 (...) -DPy_BUILD_CORE_BUILTIN -c ../../Modules/_testcapi/vectorcall_limited.c -o Modules/_testcapi/vectorcall_limited.o

In file included from ../../Modules/_testcapi/vectorcall_limited.c:2:
In file included from ../../Modules/_testcapi/parts.h:7:
In file included from ../../Include/Python.h:44:
../../Include/pyport.h:52:4: error: "Py_LIMITED_API is not compatible with Py_BUILD_CORE"
#  error "Py_LIMITED_API is not compatible with Py_BUILD_CORE"
   ^
1 error generated.
```

I'm not sure how ``-DPy_BUILD_CORE_BUILTIN`` landed in the command building ``Modules/_testcapi/vectorcall_limited.c``.

I don't think that it's correct that ``vectorcall_limited.c`` which tests the limited C API on purpose it built with ``Py_BUILD_CORE``.

On my Linux machine, Makefile contains:

```
Modules/_testcapi/vectorcall_limited.o: $(srcdir)/Modules/_testcapi/vectorcall_limited.c $(MODULE__TESTCAPI_DEPS) $(MODULE_DEPS_SHARED) $(PYTHON_HEADERS); $(CC) $(MODULE__TESTCAPI_CFLAGS) $(PY_STDMODULE_CFLAGS) $(CCSHARED) -c $(srcdir)/Modules/_testcapi/vectorcall_limited.c -o Modules/_testcapi/vectorcall_limited.o
```

So it gets two groups of compiler flags:

* ``MODULE__TESTCAPI_CFLAGS``: not defined (empty)
* ``PY_STDMODULE_CFLAGS``: ``-fno-strict-overflow -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -g -Og -Wall -O0   -std=c11 -Werror=implicit-function-declaration -fvisibility=hidden  -I./Include/internal  -I. -I./Include``

I don't see ``-DPy_BUILD_CORE_BUILTIN`` here.

**Comment by vstinner** [comment]
> Oh. This change broke WASM buildbots.

I created issue #109723 to track this issue.

**Pull request merged**